mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,323 @@
+            enable_backpressure: true,
+            initial_read_window_size: PART_SIZE,
+            ..Default::default()
+        };
+        let client = MockClient::new(config);
+        (
+            client.clone(),
+            ExpressDataCache::new(
+                bucket,
+                client,
+                "unique source description",
+                BLOCK_SIZE,
+                EXPRESS_CACHE_MAX_OBJECT_SIZE,
+            ),
+        )
+    }
+
+    #[tokio::test]
+    async fn test_put_in_both_caches() {
+        let disk_cache = default_disk_cache();
+        let (client, express_cache) = default_express_cache();
+        let runtime = ThreadPool::builder().pool_size(1).create().unwrap();
+        let cache = MultilevelDataCache::new(disk_cache, express_cache, runtime);
+
+        let data = ChecksummedBytes::new("Foo".into());
+        let object_size = data.len();
+        let cache_key = ObjectId::new("a".into(), ETag::for_tests());
+
+        // put in both caches
+        cache
+            .put_block(cache_key.clone(), 0, 0, data.clone(), object_size)
+            .await
+            .expect("put should succeed");
+
+        // check it was put to express
+        assert_eq!(client.objects_number(), 1);
+
+        // check it was put to local
+        client.remove_all_objects();
+        let entry = cache
+            .get_block(&cache_key, 0, 0, object_size)
+            .await
+            .expect("cache should be accessible")
+            .expect("cache entry should be returned");
+        assert_eq!(
+            data, entry,
+            "cache entry returned should match original bytes after put"
+        );
+    }
+
+    #[tokio::test]
+    async fn test_put_from_express_to_local() {
+        let disk_cache = default_disk_cache();
+        let (client, express_cache) = default_express_cache();
+
+        let data = ChecksummedBytes::new("Foo".into());
+        let object_size = data.len();
+        let cache_key = ObjectId::new("a".into(), ETag::for_tests());
+        express_cache
+            .put_block(cache_key.clone(), 0, 0, data.clone(), object_size)
+            .await
+            .expect("put should succeed");
+
+        let runtime = ThreadPool::builder().pool_size(1).create().unwrap();
+        let cache = MultilevelDataCache::new(disk_cache, express_cache, runtime);
+
+        // get from express, put entry in the local cache
+        let entry = cache
+            .get_block(&cache_key, 0, 0, object_size)
+            .await
+            .expect("cache should be accessible")
+            .expect("cache entry should be returned");
+        assert_eq!(
+            data, entry,
+            "cache entry returned should match original bytes after put"
+        );
+
+        // delete entry from express
+        client.remove_all_objects();
+
+        // get entry from the local cache (with retries as it is async)
+        let mut retries = 10;
+        let entry = loop {
```

Comment:
Why a loop here? Are we expecting it to fail sometimes?

Reply:
yep, cache population is async so we don't really know how along after `get_block` completed the data'll be in cache

Reply:
Could we extend this so `get_block` or whatever returns a token that can be awaited/used to know when the cache has been written to?

Reply:
You mean something like this?
```rust
    async fn get_block(
        &self,
        cache_key: &ObjectId,
        block_idx: BlockIndex,
        block_offset: u64,
    ) -> DataCacheResult<(Option<ChecksummedBytes>, Option<RemoteHandle<..>>)>;
```

This doesn't look like a good fit for this API. Consider that there are multiple implementations of `get_block` and the second returned value will only make sense in the context of multilevel cache.

NB: there is a way to [spawn a task with handle](https://docs.rs/futures/latest/futures/task/trait.SpawnExt.html#method.spawn_with_handle), but the other tricky bit is that the task is canceled when the handle is dropped, so we'll need to hold this handle in the production code just for the sake of testing